experimental/python/wheel.bzl

```diff
@@ -0,0 +1,159 @@
+# limitations under the License.
+
+"""Rules for building wheels."""
+
+def _py_wheel_impl(ctx):
+    outfile = ctx.actions.declare_file("-".join([
+        ctx.attr.wheel_name,
+        ctx.attr.version,
+        ctx.attr.python,
+        ctx.attr.abi,
+        ctx.attr.platform,
+    ]) + ".whl")
+
+    inputs = depset(
+        transitive = [dep[DefaultInfo].data_runfiles.files for dep in ctx.attr.deps] +
+                     [dep[DefaultInfo].default_runfiles.files for dep in ctx.attr.deps],
+    )
+    arguments = [
+        "--name",
+        ctx.attr.wheel_name,
+        "--version",
+        ctx.attr.version,
+        "--python",
+        ctx.attr.python,
+        "--abi",
+        ctx.attr.abi,
+        "--platform",
+        ctx.attr.platform,
+        "--out",
+        outfile.path,
+    ]
+    arguments.extend(["--restrict_package=%s" % p for p in ctx.attr.packages])
+    for input_file in inputs.to_list():
+        arguments.append("--input_file")
+        arguments.append("%s;%s" % (input_file.short_path, input_file.path))
+
+    extra_headers = []
+    if ctx.attr.author:
+        extra_headers.append("Author: %s" % ctx.attr.author)
+    if ctx.attr.author_email:
+        extra_headers.append("Author-email: %s" % ctx.attr.author_email)
+    if ctx.attr.homepage:
+        extra_headers.append("Home-page: %s" % ctx.attr.homepage)
+    if ctx.attr.license:
+        extra_headers.append("License: %s" % ctx.attr.license)
+
+    for h in extra_headers:
+        arguments.append("--header")
+        arguments.append(h)
+
+    for c in ctx.attr.classifiers:
+        arguments.append("--classifier")
+        arguments.append(c)
+
+    for r in ctx.attr.requires:
+        arguments.append("--requires")
+        arguments.append(r)
+
+    for option, requirements in ctx.attr.extra_requires.items():
+        for r in requirements:
+            arguments.append("--extra_requires")
+            arguments.append(r + ";" + option)
+
+    for name, ref in ctx.attr.console_scripts.items():
+        arguments.append("--console_script")
+        arguments.append(name + " = " + ref)
+
+    if ctx.attr.description_file:
+        description_files = ctx.attr.description_file.files.to_list()
+        arguments.append("--description_file")
+        arguments.append(description_files[0].path)
+        inputs = inputs.union(ctx.attr.description_file.files)
+
+    ctx.actions.run(
+        inputs = inputs,
+        outputs = [outfile],
+        arguments = arguments,
+        executable = ctx.executable._wheelmaker,
+        progress_message = "Building wheel",
+    )
+    return [DefaultInfo(
+        files = depset([outfile]),
+        data_runfiles = ctx.runfiles(files = [outfile]),
+    )]
+
+py_wheel = rule(
+    implementation = _py_wheel_impl,
+    attrs = {
+        "deps": attr.label_list(
+            doc = "Dependencies of this package, e.g. py_library rules",
+        ),
+        "packages": attr.string_list(
```

Comment:
It seems like this filtering operates at the level of Bazel packages rather than Python packages. The two might not be the same, for instance if a `py_library` uses the `imports` field to add a subpackage as a new root to the `PYTHONPATH`. Maybe change this to a list of labels, or a list of Bazel package prefix strings?

Reply:
On further reflection, I think it might be better to completely divorce the policy decision of *what* files to package from the mechanism that does this packaging. This makes it easier to change the behavior if users have different needs -- or if everyone has the same need but it's different from what we think is the best mechanism right now. :)

This could be done by replacing the `deps` attribute with a `files` attribute. Instead of expanding the transitive runfiles of each item in `deps`, we would just union the direct files of each item in `files` (that is, take `ctx.files.files`). Then some other rule could be used to aggregate the runfiles of some given deps and filter them based on package prefix. Or not. Up to the user. (This does mean the filtering would be moved from the Python tool into the new .bzl rule. I believe it may still be possible to filter without flattening the depset by using `Args.add_all`'s [`map_each`](https://g3doc.corp.google.com/devtools/blaze/rules/g3doc/lib/Args.html?cl=head#add_all.map_each) parameter.)

I was chatting with @aiuto about this but I don't think we're 100% resolved. I'm out Wednesday but back Thursday.

Reply:
PTAL: I've separated selection of files and packaging. Right now there's just one selection mechanism, py_package which selects a package recursively.

Let me know if that's what you have envisioned - if so I'll work on documentation and tests.

Args.map_each (btw, the link you provided seems to be google internal resource) only works at Args conversion time, which is much later. Since I guess we want to allow multiple source rules (be it Fileset, py_package or something else) for one py_wheel rule, we need to use depsets for merging; alas, depsets don't support removal by design (https://docs.bazel.build/versions/master/skylark/depsets.html#description-and-operations). While I can avoid to_list when build args, I don't see how to do this in py_package rule.

Reply:
Added example usages with both traversal of dependencies (via new py_package rule) and including just given py_library, without dependencies.

Reply:
Thanks, I think this addresses my concern. Please add / factor-out a test for the new filtering rule (if it's not already in there and I missed it).

Reply:
Added a test